### DISCUSSION TITLE
Problems with Invalid PPM Files on Windows

### DISCUSSION THREAD

**Discussion by hollasch** [RayTracing/raytracing.github.io#1114] [Debugging]
(See issue #636 for more context.)

Several readers have reported problems with invalid PPM files generated from their raytracers on Windows. The short answer is that there are several ways to have text files output as UTF-16, or even UTF-16 with a leading BOM, which is unrecognized by most PPM readers.

If your PPM file shows as invalid, view the hex contents of the file and verify that the first bytes are 0x50, 0x30 ("P3"), and the next two should be either 0x0d 0x0a (CRLF), or just 0x0a (LF). If not, your build or run environment is probably outputting Unicode text as either UTF-16 or UTF-16 with a leading BOM. The BOM will appear as either 0xFE, 0xFF, or 0xFF, 0xFE. For example, one reader was generating the following bytes in their PPM output:

    ff fe  50 00  33 00 ...
    -----  -----  -----
     BOM     P      3

In either event, you'll need to figure out how to configure your project to output ASCII or UTF-8.

See issue #636 for discussion on this topic. Both readers reported that the problem manifested when using PowerShell. To resolve it, there were two options:

- Run the programs from a CMD shell
- Run from PowerShell like so: `PS: C:\raytracing\main\Release> main.exe | set-content image.ppm -encoding String`

**Comment by LiamGeorge1999** [comment]
I'm fairly sure that 0x30 is actually "0", and that 0x33 is "3", as shown in hexed.it here:
![image](https://github.com/RayTracing/raytracing.github.io/assets/77801642/c64b8cdd-6d3d-4081-87a9-6768a5e1a68b)

**Reply by hollasch** [reply]
Yes, my mistake. Hex 33 is the character '3'.

**Comment by ayaanmaan** [comment]
PS might prompt you to write ".\main.exe" in place of "main.exe". After 12 hours I finally realized I was writing it with the double quotes. Please don't be like me and take things too literally.
`.\main.exe | set-content image.ppm -encoding String` works.

**Reply by NSYSUTim** [reply]
Same as me. But why we can't write the file name with double quotes if it has a space in it? like this:
"ray tracing".exe
It works well if I change the name into ray_tracing.exe and eliminate the double quote in the PowerShell.
(Thanks for your comment by the way.)

**Comment by North-Mi** [comment]
there maybe some encoding problem. if you find the file is right but can\`t open the .ppm picture, you might to check the file encoding, I has this problem and use the windows api  `SetConsoleOutputCP(CP_UTF8);` to solve it.

**Comment by conniedeng303** [comment]
noob question but was anyone able to solve this issue with another method? I followed the post, as well as all the comments

**Reply by kryptodogg** [reply]
I just opened the file in Notepad and saved the .ppm with UTF-8 encoding and it opened fine.

I know this is a while after the fact, but maybe it'll save someone some headache.

**Comment by Ooooscar** [comment]
I'm using VSCode with MinGW-w64 and GCC.

Just in case that anyone also has no idea about Windows API like me,
I got the very first example working as follows (thanks for the help, ChatGPT):

```c++
#include <iostream>
#include <fstream>

#ifdef _WIN32
#include <windows.h>
#endif

int main() {

#ifdef _WIN32
    SetConsoleOutputCP(CP_UTF8);
#endif

    std::ofstream output_file{"image.ppm", std::ios::out | std::ios::binary};
    if (!output_file) {
        std::cerr << "Error: Could not open file for writing.\n";
        return EXIT_FAILURE;
    }

    // Image

    int image_width = 256;
    int image_height = 256;

    // Render

    output_file << "P3\n" << image_width << ' ' << image_height << "\n255\n";

    for (int j = 0; j < image_height; j++) {
        for (int i = 0; i < image_width; i++) {
            auto r = double(i) / (image_width-1);
            auto g = double(j) / (image_height-1);
            auto b = 0.0;

            int ir = int(255.999 * r);
            int ig = int(255.999 * g);
            int ib = int(255.999 * b);

            output_file << ir << ' ' << ig << ' ' << ib << '\n';
        }
    }
    
    output_file.flush();
    output_file.close();

    return EXIT_SUCCESS;
}
```

**Comment by Raz1789** [comment]
If anyone is using Visual Studio Code and is apprehensive about adding more code, you can switch the terminal to use Bash (which should be installed if you have Git) by clicking the "+" dropdown where it says "PowerShell" in the VS Code terminal. This will build it in the proper format. Hope it helps! Also, if it helps, I use IrfanView to view the actual PPM files on Windows.
